clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/NettyP2PBlobTransferManager.java

```diff
@@ -39,31 +48,124 @@ public void start() throws Exception {
-    CompletionStage<InputStream> inputStream;
+    // error case 1: no peers are found for the requested blob
     BlobPeersDiscoveryResponse response = peerFinder.discoverBlobPeers(storeName, version, partition);
     if (response == null || response.isError()) {
       throw new VenicePeersNotFoundException("Failed to obtain the peers for the requested blob");
     }
+
     List<String> discoverPeers = response.getDiscoveryResult();
     if (discoverPeers == null || discoverPeers.isEmpty()) {
       throw new VenicePeersNotFoundException("No peers found for the requested blob");
     }
+    LOGGER
+        .info("Discovered peers {} for store {} version {} partition {}", discoverPeers, storeName, version, partition);
+
+    Instant startTime = Instant.now();
     for (String peer: discoverPeers) {
-      try {
-        // TODO: add some retry logic or strategy to choose the peers differently in case of failure
-        // instanceName comes as a format of <hostName>_<applicationPort>
-        String chosenHost = peer.split("_")[0];
-        LOGGER.info("Chosen host: {}", chosenHost);
-        inputStream = nettyClient.get(chosenHost, storeName, version, partition);
-        return inputStream;
-      } catch (Exception e) {
-        LOGGER.warn("Failed to connect to peer: {}", peer, e);
+      int retryCount = 0;
+      while (retryCount < MAX_RETRIES_FOR_BLOB_TRANSFER_PER_HOST) {
+        try {
+          // instanceName comes as a format of <hostName>_<applicationPort>
+          String chosenHost = peer.split("_")[0];
+          LOGGER.info("Attempt {} to connect to host: {}", retryCount + 1, chosenHost);
+          CompletionStage<InputStream> inputStreamStage = nettyClient.get(chosenHost, storeName, version, partition);
+          CompletableFuture<InputStream> inputStreamFuture = inputStreamStage.toCompletableFuture();
+
+          InputStream inputStream = inputStreamFuture.get(MAX_TIMEOUT_FOR_BLOB_TRANSFER_IN_MIN, TimeUnit.MINUTES);
+          LOGGER.info(
+              "Successfully fetched blob from peer {} for store {} partition {} version {} in {} seconds",
+              peer,
+              storeName,
+              partition,
+              version,
+              Duration.between(startTime, Instant.now()).getSeconds());
+          return CompletableFuture.completedFuture(inputStream);
+        } catch (Exception e) {
+          if (e.getCause() instanceof VenicePeersCannotConnectException) {
+            // error case 2: failed to connect to the peer,
+            // solution: retry connecting to the peer again up to MAX_RETRIES_FOR_BLOB_TRANSFER_PER_HOST times
+            LOGGER.warn(
+                "Get error when connect to peer: {} for store {} partition {}, retrying {}/{}",
+                peer,
+                storeName,
+                partition,
+                retryCount + 1,
+                MAX_RETRIES_FOR_BLOB_TRANSFER_PER_HOST,
+                e);
+            retryCount++;
```

Comment:
It because for single host connect, this is reasonable that there could be connect issue, so we can retry connect, but for other issue, [file transfer error, file not existed error], we can directly move on.